### ISSUE TITLE
Cannot make query to db within migration

### ISSUE THREAD

**Issue by frostmark** [rails/rails#49896]
Hello! I'm upgrading rails from 7.0 to 7.1 and got some problems with migrations

### Steps to reproduce
When I'm trying to get records from db within migration it freezes and does nothing. (I guess something related with transaction and isolation level here, but not sure)
```ruby
# frozen_string_literal: true

class SomeMigration < ActiveRecord::Migration[7.1]
  def change
    add_reference :files, :asset, foreign_key: true

    File.all.each do |f| # <- Freezes here and does nothing
      f.update(asset_id: f.folder.asset_id)
    end
  end
end

```

### Expected behavior
Can make query within migration file

### Actual behavior
Query doesn't work

### System configuration
Rails 7.1.1

ruby 3.2.2

**Comment by fatkodima** [comment]
Which database do you use?

**Comment by frostmark** [comment]
@paulreece postgres 14.9

**Comment by frostmark** [comment]
@fatkodima postgres 14.9

**Comment by frostmark** [comment]
> Hey there. So in general it's recommended to not reference the ActiveRecord Model in the migrations. A better approach would be to add this to your `db/seed.rb` file, then run the migration above and seed your DB:
> 
> ```ruby
> File.all.each do |f| 
>       f.update(asset_id: f.folder.asset_id)
>     end
> ```

yeah, it's just old migration where we updated some data within migration. Just decided re-run all migrations after upgrade rails and got the problem. In rails 7.0 is no this issue

**Comment by fatkodima** [comment]
Ok, so while I agree that it is not a good idea to write data migrations like this, there shouldn't be a freeze.

The whole migration is run within transaction, `add_reference` creates an exclusive lock on the `files` and `assets` tables, but `File.all` issues `SELECT` on the locked table, but within the same transaction, so should work.

Can you reproduce this within a sample rails app? If yes, can you share its code?

**Comment by matthewd** [comment]
> A better approach would be to add this to your db/seed.rb file

Data transformation is really not appropriate for seeds, IMO. Depending on application-defined models is unwise, but this update belongs in a migration, whether as an `execute "UPDATE .."`, or a migration-local temporary AR model.

> `File`

I assume we're all just assuming that's not the real class name.

> The whole migration is run within transaction, add_reference creates an exclusive lock on the files and assets tables, but File.all issues SELECT on the locked table, but within the same transaction, so should work.

Is multiple database support involved here? That could break the above same-transaction assumption (when combined with use of the real model).

**Comment by paulreece** [comment]
>  `File`
> I assume we're all just assuming that's not the real class name.

Given that the migration is called `SomeMigration` I think that's a safe assumption. 

We need more context and a repo script if possible.

**Comment by frostmark** [comment]
Hi guys! Thanks for your helping!

Here is repo where you can reproduce it. (I have configured it like my real application)
https://github.com/frostmark/demo_rails/tree/main

> Is multiple database support involved here? That could break the above same-transaction assumption (when combined with use of the real model).

Yes I'm using multiple databases (sharding) and it works fine on 7.0

**Comment by fatkodima** [comment]
I bisected this to https://github.com/rails/rails/pull/48353.
cc @eileencodes

**Comment by eileencodes** [comment]
I'm not sure there's an easy solution to this in Rails. The problem is the migration opens its own connection and transaction using role `:writing` and shard `:default`. But the application doesn't have a `:default` shard, it has `:shard_one` and `:shard_two` so the transaction is a different connection context than the queries run by the migration.

One possible work around is to mark these migrations as not DDL migrations with `disable_ddl_transactions!` in the top of the migration file. Another option is to set the application default shard to `:shard_one` with `config.active_record.default_shard = :shard_one` in an initializer.

I'll keep working on trying to find a proper fix that doesn't require intervention on the application side, but I think this is a non-trivial bug to fix.

**Comment by frostmark** [comment]
@eileencodes yeah thanks! In my case I just renamed `:primary` to `:default` and it resolved the problem. BTW maybe would be nice to update this within documentation? 

looks a bit misleading
<img width="638" alt="image" src="https://github.com/rails/rails/assets/4071096/99f476fd-84a3-4566-a719-e86010db6f4d">

https://guides.rubyonrails.org/active_record_multiple_databases.html#setting-up-your-application

**Comment by eileencodes** [comment]
I reopened this because it is still a bug, just not one with an easy fix. There's a feature I was planning to implement that I think will resolve this and many other issues we have with multi-db and when connections are established.
